File: src/utils/recipeParser.ts
export interface ParsedRecipe {
  id: string;
  name: string;
  region: string;
  category: string;
  ingredients: string[];
  steps: string[];
  image?: string;
  cookTime?: string;
  servings?: string;
  tips?: string[];
}

/**
 * Parse recipe text content from the uploaded recipe files
 */
export function parseRecipeText(content: string, recipeName: string): ParsedRecipe {
  const lines = content.split('\n').map(line => line.trim()).filter(Boolean);
  
  const recipe: ParsedRecipe = {
    id: recipeName.toLowerCase().replace(/\s+/g, '-'),
    name: '',
    region: '',
    category: '',
    ingredients: [],
    steps: []
  };
  
  let currentSection = '';
  
  lines.forEach(line => {
    if (line.startsWith('Recipe Name:')) {
      recipe.name = line.replace('Recipe Name:', '').trim();
    } else if (line.startsWith('Region Style:')) {
      recipe.region = line.replace('Region Style:', '').trim();
    } else if (line.startsWith('Category:')) {
      recipe.category = line.replace('Category:', '').trim();
    } else if (line === 'Ingredients:') {
      currentSection = 'ingredients';
    } else if (line === 'Steps:') {
      currentSection = 'steps';
    } else if (line.startsWith('-') && currentSection === 'ingredients') {
      recipe.ingredients.push(line.replace('-', '').trim());
    } else if (line.match(/^\d+\./) && currentSection === 'steps') {
      recipe.steps.push(line.replace(/^\d+\.\s*/, '').trim());
    }
  });
  
  return recipe;
}

/**
 * Sample recipes extracted from the Tamil Nadu recipe collection
 * These represent different regions and categories
 */
export const tamilNaduRecipes: ParsedRecipe[] = [
  {
    id: "kongu-nadu-sambar",
    name: "Sambar",
    region: "Kongu Nadu",
    category: "Vegetarian",
    ingredients: [
      "Traditional Kongu Nadu spices and ingredients",
      "Turmeric, chilli, coriander, curry leaves",
      "Dal, vegetables"
    ],
    steps: [
      "Prepare fresh ingredients",
      "Start with regional-style tempering",
      "Add main ingredient and spices",
      "Cook until aroma develops",
      "Add water or coconut milk as needed",
      "Simmer and garnish with coriander"
    ]
  },
  {
    id: "chola-nadu-fish-kuzhambu",
    name: "Fish Kuzhambu",
    region: "Chola Nadu",
    category: "Non-Vegetarian",
    ingredients: [
      "Traditional Chola Nadu spices and ingredients",
      "Turmeric, chilli, coriander, curry leaves",
      "Fresh fish, tamarind"
    ],
    steps: [
      "Prepare fresh ingredients",
      "Start with regional-style tempering",
      "Add main ingredient and spices",
      "Cook until aroma develops",
      "Add water or coconut milk as needed",
      "Simmer and garnish with coriander"
    ]
  }
];

/**
 * Parse recipe from AI chat response text
 */
export function parseRecipeFromAIResponse(text: string): ParsedRecipe | null {
  try {
    const lines = text.split('\n').filter(line => line.trim());
    
    // Extract recipe name (usually first line or after a heading)
    let name = lines[0]?.trim() || 'Recipe';
    
    // Remove common prefixes
    name = name.replace(/^(Recipe:|Recipe Name:|\[|\])/gi, '').trim();
    
    // Extract ingredients
    const ingredientsStart = lines.findIndex(line => 
      /^Ingredients:/i.test(line.trim())
    );
    const stepsStart = lines.findIndex(line => 
      /^(Steps:|Instructions:|Step-by-step|Method:)/i.test(line.trim())
    );
    
    const ingredients: string[] = [];
    if (ingredientsStart !== -1 && stepsStart !== -1) {
      for (let i = ingredientsStart + 1; i < stepsStart; i++) {
        const ingredient = lines[i]?.trim().replace(/^[-•*]\s*/, '');
        if (ingredient && !ingredient.match(/^(Steps|Instructions|Method):/i)) {
          ingredients.push(ingredient);
        }
      }
    }
    
    // Extract steps
    const steps: string[] = [];
    if (stepsStart !== -1) {
      for (let i = stepsStart + 1; i < lines.length; i++) {
        const line = lines[i]?.trim();
        if (!line) continue;
        
        // Stop if we hit other sections
        if (/^(Cooking Time:|Serves:|Tips:|Notes:)/i.test(line)) break;
        
        // Remove step numbering and bullet points
        const step = line.replace(/^(\d+\.|[-•*])\s*/, '').trim();
        if (step && step.length > 5) {
          steps.push(step);
        }
      }
    }
    
    // Extract metadata
    const cookTimeMatch = text.match(/Cooking Time:\s*(.+?)(?:\n|$)/i);
    const servingsMatch = text.match(/Serves?:\s*(.+?)(?:\n|$)/i);
    
    // Extract tips
    const tipsStart = lines.findIndex(line => /^Tips:/i.test(line.trim()));
    const tips: string[] = [];
    if (tipsStart !== -1) {
      for (let i = tipsStart + 1; i < lines.length; i++) {
        const tip = lines[i]?.trim().replace(/^[-•*]\s*/, '');
        if (tip && tip.length > 5) tips.push(tip);
      }
    }
    
    if (steps.length === 0) {
      return null;
    }
    
    return {
      id: name.toLowerCase().replace(/\s+/g, '-'),
      name,
      region: '',
      category: '',
      ingredients,
      steps,
      cookTime: cookTimeMatch?.[1]?.trim(),
      servings: servingsMatch?.[1]?.trim(),
      tips: tips.length > 0 ? tips : undefined,
    };
  } catch (error) {
    console.error('Error parsing recipe:', error);
    return null;
  }
}

File: app/utils/recipe_parser.py
"""
Recipe parsing utilities
"""
import re
from typing import Optional, Dict, List


class ParsedRecipe:
    """Parsed recipe data structure"""
    def __init__(self):
        self.id: str = ""
        self.name: str = ""
        self.region: str = ""
        self.category: str = ""
        self.ingredients: List[str] = []
        self.steps: List[str] = []
        self.image: Optional[str] = None
        self.cook_time: Optional[str] = None
        self.servings: Optional[str] = None
        self.tips: List[str] = []

    def to_dict(self) -> Dict:
        """Convert to dictionary"""
        return {
            "id": self.id,
            "name": self.name,
            "region": self.region,
            "category": self.category,
            "ingredients": self.ingredients,
            "steps": self.steps,
            "image": self.image,
            "cookTime": self.cook_time,
            "servings": self.servings,
            "tips": self.tips if self.tips else None
        }


def parse_recipe_text(content: str, recipe_name: str) -> ParsedRecipe:
    """Parse recipe text content from uploaded recipe files"""
    lines = [line.strip() for line in content.split('\n') if line.strip()]
    
    recipe = ParsedRecipe()
    recipe.id = recipe_name.lower().replace(' ', '-')
    current_section = ''
    
    for line in lines:
        if line.startswith('Recipe Name:'):
            recipe.name = line.replace('Recipe Name:', '').strip()
        elif line.startswith('Region Style:'):
            recipe.region = line.replace('Region Style:', '').strip()
        elif line.startswith('Category:'):
            recipe.category = line.replace('Category:', '').strip()
        elif line == 'Ingredients:':
            current_section = 'ingredients'
        elif line == 'Steps:':
            current_section = 'steps'
        elif line.startswith('-') and current_section == 'ingredients':
            recipe.ingredients.append(line.replace('-', '', 1).strip())
        elif re.match(r'^\d+\.', line) and current_section == 'steps':
            recipe.steps.append(re.sub(r'^\d+\.\s*', '', line).strip())
    
    return recipe


def parse_recipe_from_ai_response(text: str) -> Optional[ParsedRecipe]:
    """Parse recipe from AI chat response text"""
    try:
        lines = [line.strip() for line in text.split('\n') if line.strip()]
        
        if not lines:
            return None
        
        recipe = ParsedRecipe()
        
        # Extract recipe name (usually first line)
        name = lines[0] if lines else 'Recipe'
        name = re.sub(r'^(Recipe:|Recipe Name:|\[|\])', '', name, flags=re.IGNORECASE).strip()
        recipe.name = name
        recipe.id = name.lower().replace(' ', '-')
        
        # Find sections
        ingredients_start = next((i for i, line in enumerate(lines) 
                                 if re.match(r'^Ingredients:', line, re.IGNORECASE)), -1)
        steps_start = next((i for i, line in enumerate(lines) 
                           if re.match(r'^(Steps:|Instructions:|Step-by-step|Method:)', line, re.IGNORECASE)), -1)
        
        # Extract ingredients
        if ingredients_start != -1 and steps_start != -1:
            for i in range(ingredients_start + 1, steps_start):
                ingredient = re.sub(r'^[-•*]\s*', '', lines[i]).strip()
                if ingredient and not re.match(r'^(Steps|Instructions|Method):', ingredient, re.IGNORECASE):
                    recipe.ingredients.append(ingredient)
        
        # Extract steps
        if steps_start != -1:
            for i in range(steps_start + 1, len(lines)):
                line = lines[i]
                if not line:
                    continue
                
                # Stop if we hit other sections
                if re.match(r'^(Cooking Time:|Serves:|Tips:|Notes:)', line, re.IGNORECASE):
                    break
                
                # Remove step numbering and bullet points
                step = re.sub(r'^(\d+\.|[-•*])\s*', '', line).strip()
                if step and len(step) > 5:
                    recipe.steps.append(step)
        
        # Extract metadata
        cook_time_match = re.search(r'Cooking Time:\s*(.+?)(?:\n|$)', text, re.IGNORECASE)
        servings_match = re.search(r'Serves?:\s*(.+?)(?:\n|$)', text, re.IGNORECASE)
        
        if cook_time_match:
            recipe.cook_time = cook_time_match.group(1).strip()
        if servings_match:
            recipe.servings = servings_match.group(1).strip()
        
        # Extract tips
        tips_start = next((i for i, line in enumerate(lines) 
                          if re.match(r'^Tips:', line, re.IGNORECASE)), -1)
        if tips_start != -1:
            for i in range(tips_start + 1, len(lines)):
                tip = re.sub(r'^[-•*]\s*', '', lines[i]).strip()
                if tip and len(tip) > 5:
                    recipe.tips.append(tip)
        
        if not recipe.steps:
            return None
        
        return recipe
    except Exception as e:
        print(f'Error parsing recipe: {e}')
        return None

